packages/runtime/container-runtime/src/garbageCollection.ts

```diff
@@ -878,44 +887,52 @@ export class GarbageCollector implements IGarbageCollector {
             0x2b7, /* "Can't validate correctness without GC data from last run" */
         );
 
+        const getDataStoreReferencesFromParentToChild = (nodeId: string, references: string[]) =>
+            references.filter((reference) => !nodeId.startsWith(reference) && isDataStoreNode(reference));
+
         // Get a list of all the outbound routes (or references) in the current GC data.
         const currentReferences: string[] = [];
-        for (const [nodeId, outboundRoutes] of Object.entries(currentGCData.gcNodes)) {
+        for (const [nodeId, references] of Object.entries(currentGCData.gcNodes)) {
             /**
              * Remove routes from a child node to its parent which is added implicitly by the runtime. For instance,
-             * each adds its data store as an outbound route to mark it as referenced if the DDS is referenced.
-             * We won't get any explicit notification for these references so they must be removed before validation.
+             * each DDS adds its data store as an outbound route to mark it as referenced if the DDS is referenced.
+             * We won't get any notifications for these new references so they must be removed before validation.
              */
-            const explicitRoutes = outboundRoutes.filter((route) => !nodeId.startsWith(route));
-            currentReferences.push(...explicitRoutes);
+            const currentDataStoreReferences = getDataStoreReferencesFromParentToChild(nodeId, references);
+            currentReferences.push(...currentDataStoreReferences);
         }
 
-        // Get a list of outbound routes (or references) from the last run's GC data plus references added since the
-        // last run that were notified via `referenceAdded`.
-        const explicitReferences: string[] = [];
-        for (const [, outboundRoutes] of Object.entries(this.gcDataFromLastRun.gcNodes)) {
-            explicitReferences.push(...outboundRoutes);
+        // Get a list of references (or outbound routes) from the previous run's GC data plus new references added since
+        // the last run that were notified via `addedOutboundReference`.
+        const previousAndNewReferences: string[] = [];
+        const previousGraph = Object.entries(this.gcDataFromLastRun.gcNodes);
```

Comment:
We should probably rename `gcDataFromLastRun` -> `previousGCData` or `previousRunGCData` to make it clearer.